### PR TITLE
AST validate that a contract does not have a function with contract name

### PR DESCRIPTION
**Pull request by ferranbt**
This PR adds an AST validation check similar to [this](https://github.com/ethereum/solidity/blob/df01dcc7b759f53ba45811979d3286562eab7846/libsolidity/analysis/SyntaxChecker.cpp#L385) which verifies that a function does not have the same name as the contract.

Note, eventually this can be merged into a specific AST contract level checker.

### CODE REVIEW

File: crates/sema/src/ast_passes.rs [hunk 1]
```diff
@@ -109,6 +109,31 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
...
+        for item in body.iter() {
+            if let ast::ItemKind::Function(ast::ItemFunction { kind: _, header, body: _ }) =
+                &item.kind
+            {
+                if let Some(func_name) = header.name {
+                    if func_name.as_str() == contract_name {
+                        self.dcx()
+                            .err(
+                                "Functions are not allowed to have the same name as the contract.
+        If you intend this to be a constructor, use \"constructor(...) { ... }\" to define it.",
```
Comment (onbjerg) [inline crates/sema/src/ast_passes.rs]: we should lowercase the start of the error and move the "if you intend..." into a `note(..)` call instead

File: crates/sema/src/ast_passes.rs [hunk 2]
```diff
@@ -123,6 +123,32 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
...
+
+        for item in body.iter() {
+            if let ast::ItemKind::Function(ast::ItemFunction { kind: _, header, body: _ }) =
+                &item.kind
+            {
+                if let Some(func_name) = header.name {
+                    if func_name.as_str() == contract_name {
+                        self.dcx()
+                            .err("functions are not allowed to have the same name as the contract")
+                            .note("if you intend this to be a constructor, use \"constructor(...) { ... }\" to define it")
```
Comment (DaniPopes) [inline crates/sema/src/ast_passes.rs]:
```suggestion
                            .note("if you intend this to be a constructor, use `constructor(...) { ... }` to define it")
```

**Review by onbjerg — APPROVED** [review]
lgtm

File: crates/sema/src/ast_passes.rs [hunk 3]
```diff
@@ -123,6 +123,32 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
...
+    ) -> ControlFlow<Self::BreakValue> {
+        let ast::ItemContract { kind: _, name, bases: _, body } = contract;
+        let contract_name = name.as_str();
+
+        for item in body.iter() {
+            if let ast::ItemKind::Function(ast::ItemFunction { kind: _, header, body: _ }) =
+                &item.kind
+            {
+                if let Some(func_name) = header.name {
+                    if func_name.as_str() == contract_name {
```
Comment (DaniPopes) [inline crates/sema/src/ast_passes.rs]:
```suggestion
                    if func_name == contract.name {
```

Compare Symbols not strings, it has the same result but faster, which is the point of Symbol in the first place

File: crates/sema/src/ast_passes.rs [hunk 4]
```diff
@@ -123,6 +123,32 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
...
+        contract: &'ast ast::ItemContract<'ast>,
+    ) -> ControlFlow<Self::BreakValue> {
+        let ast::ItemContract { kind: _, name, bases: _, body } = contract;
+        let contract_name = name.as_str();
+
+        for item in body.iter() {
+            if let ast::ItemKind::Function(ast::ItemFunction { kind: _, header, body: _ }) =
+                &item.kind
+            {
+                if let Some(func_name) = header.name {
```
Comment (DaniPopes) [inline crates/sema/src/ast_passes.rs]: we need to also check that kind is Function

File: crates/sema/src/ast_passes.rs [hunk 5]
```diff
@@ -123,6 +123,32 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
...
+    fn visit_item_contract(
+        &mut self,
+        contract: &'ast ast::ItemContract<'ast>,
+    ) -> ControlFlow<Self::BreakValue> {
+        let ast::ItemContract { kind: _, name: _, bases: _, body } = contract;
+
+        for item in body.iter() {
+            if let ast::ItemKind::Function(ast::ItemFunction { kind, header, body: _ }) = &item.kind
+            {
+                if *kind == ast::FunctionKind::Function {
```
Comment (ferranbt) [inline crates/sema/src/ast_passes.rs:135]: Unsure if there is a more idiomatic way to chain this if statements.

Comment (DaniPopes) [reply crates/sema/src/ast_passes.rs:135]: There will be when let-chains are stabilized :)

**Pull request merged**